server/src/main/java/org/apache/druid/server/coordinator/duty/KillStalePendingSegments.java

```diff
@@ -0,0 +1,139 @@
+ * <p>
+ * While cleaning up, this duty ensures that the following pending segments are
+ * retained for at least {@link #DURATION_TO_RETAIN}:
+ * <ul>
+ * <li>Pending segments created by any active task (across all datasources)</li>
+ * <li>Pending segments created by the latest completed task (across all datasources)</li>
+ * </ul>
+ */
+public class KillStalePendingSegments implements CoordinatorDuty
+{
+  private static final Logger log = new Logger(KillStalePendingSegments.class);
+  private static final Period DURATION_TO_RETAIN = new Period("P1D");
+
+  private final OverlordClient overlordClient;
+
+  @Inject
+  public KillStalePendingSegments(OverlordClient overlordClient)
+  {
+    this.overlordClient = overlordClient;
+  }
+
+  @Override
+  public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+  {
+    final Set<String> killDatasources = new HashSet<>(
+        params.getUsedSegmentsTimelinesPerDataSource().keySet()
+    );
+    killDatasources.removeAll(
+        params.getCoordinatorDynamicConfig()
+              .getDataSourcesToNotKillStalePendingSegmentsIn()
+    );
+
+    final DateTime minCreatedTime = getMinCreatedTimeToRetain();
+    for (String dataSource : killDatasources) {
+      int pendingSegmentsKilled = FutureUtils.getUnchecked(
+          overlordClient.killPendingSegments(
+              dataSource,
+              new Interval(DateTimes.MIN, minCreatedTime)
+          ),
+          true
+      );
+      if (pendingSegmentsKilled > 0) {
+        log.info(
+            "Killed [%d] pendingSegments created before [%s] for datasource[%s].",
+            pendingSegmentsKilled, minCreatedTime, dataSource
+        );
+        params.getCoordinatorStats().add(
+            Stats.Kill.PENDING_SEGMENTS,
+            RowKey.of(Dimension.DATASOURCE, dataSource),
+            pendingSegmentsKilled
+        );
+      }
+    }
+    return params;
+  }
+
+  /**
+   * Computes the minimum created time of retainable pending segments. Any pending
+   * segment created before this time is considered stale and can be safely deleted.
+   * The limit is determined to ensure that pending segments created by any active
+   * task and the latest completed task (across all datasources) are retained for
+   * at least {@link #DURATION_TO_RETAIN}.
+   */
+  private DateTime getMinCreatedTimeToRetain()
+  {
+    // Fetch the statuses of all active tasks and the latest completed task
+    // (The Overlord API returns complete tasks in descending order of created_date.)
+    final List<TaskStatusPlus> statuses = ImmutableList.copyOf(
+        FutureUtils.getUnchecked(overlordClient.taskStatuses(null, null, 1), true)
+    );
+
+    DateTime earliestActiveTaskStart = DateTimes.nowUtc();
+    DateTime latestCompletedTaskStart = null;
+    for (TaskStatusPlus status : statuses) {
+      if (status.getStatusCode() == null) {
+        // Unknown status
```

Comment:
There seems to be a change in logic here as the created time was processed in the earlier code:
`.filter(status -> status.getStatusCode() == null || !status.getStatusCode().isComplete())`

Reply:
Ah, thanks for catching this! This is why we should have some tests.

Reply:
I have made the changes but is it okay with you if I include it in a later PR. I don't want to trigger the whole CI pipeline for this.

Reply:
Yes, I'm ok with it.

Reply:
Fixed in #14891